crates/signer/src/signer.rs

```diff
@@ -0,0 +1,202 @@
+    /// The default implementation returns [`UnsupportedOperation`](Error::UnsupportedOperation).
+    #[inline]
+    fn sign_hash(&self, hash: &B256) -> Result<Signature> {
+        let _ = hash;
+        Err(Error::UnsupportedOperation(UnsupportedSignerOperation::SignHash))
+    }
+
+    /// Signs the hash.
+    ///
+    /// Asynchronous version of [`sign_hash`](Signer::sign_hash). The default implementation
+    /// delegates to the synchronous version; see its documentation for more details.
+    #[inline]
+    async fn sign_hash_async(&self, hash: &B256) -> Result<Signature> {
+        self.sign_hash(hash)
+    }
+
+    /// Signs the hash of the provided message after prefixing it, as specified in [EIP-191].
+    ///
+    /// [EIP-191]: https://eips.ethereum.org/EIPS/eip-191
+    #[inline]
+    fn sign_message(&self, message: &[u8]) -> Result<Signature> {
+        self.sign_hash(&eip191_hash_message(message))
+    }
+
+    /// Signs the hash of the provided message after prefixing it, as specified in [EIP-191].
+    ///
+    /// Asynchronous version of [`sign_message`](Signer::sign_message). The default
+    /// implementation is the same as the synchronous version; see its documentation for more
+    /// details.
+    ///
+    /// [EIP-191]: https://eips.ethereum.org/EIPS/eip-191
+    #[inline]
+    async fn sign_message_async(&self, message: &[u8]) -> Result<Signature> {
+        self.sign_hash_async(&eip191_hash_message(message)).await
+    }
+
+    /// Signs the transaction.
+    ///
+    /// The default implementation signs the [transaction's signature hash][sighash], and optionally
+    /// applies [EIP-155] to the signature if a chain ID is present.
+    ///
+    /// [sighash]: TypedTransaction::sighash
+    /// [EIP-155]: https://eips.ethereum.org/EIPS/eip-155
+    #[cfg(TODO)]
+    #[inline]
+    fn sign_transaction(&self, tx: &TypedTransaction) -> Result<Signature> {
+        self.sign_hash(&tx.sighash()).map(|mut sig| {
+            if let Some(chain_id) = tx.chain_id() {
+                sig.apply_eip155(chain_id);
+            }
+            sig
+        })
+    }
+
+    /// Signs the transaction.
+    ///
+    /// Asynchronous version of [`sign_transaction`](Signer::sign_transaction). The default
+    /// implementation is the same as the synchronous version; see its documentation for more
+    /// details.
+    #[cfg(TODO)]
+    #[inline]
+    async fn sign_transaction_async(&self, message: &TypedTransaction) -> Result<Signature> {
+        self.sign_hash_async(&tx.sighash())
+            .map(|mut sig| {
+                if let Some(chain_id) = tx.chain_id() {
+                    sig.apply_eip155(chain_id);
+                }
+                sig
+            })
+            .await
+    }
+
+    /// Encodes and signs the typed data according to [EIP-712].
+    ///
+    /// [EIP-712]: https://eips.ethereum.org/EIPS/eip-712
+    #[cfg(feature = "eip712")]
+    #[inline]
+    fn sign_typed_data<T: SolStruct + Send + Sync>(
```

Comment:
should `Send` and `Sync` and `'static` be supertraits of `SolStruct`?  is there any situation in which we allow/want `SolStructs` to have anything other than dumb data in them?

Reply:
I generally dislike slapping "send sync static" combo everywhere